src/nvim/window.c

```diff
@@ -570,6 +570,44 @@ static void cmd_with_count(char *cmd, char_u *bufp, size_t bufsize,
   }
 }
 
+void win_set_buf(Window window, Buffer buffer, bool noautocmd, Error *err)
+{
+  win_T *win = find_window_by_handle(window, err), *save_curwin = curwin;
+  buf_T *buf = find_buffer_by_handle(buffer, err);
+  tabpage_T *tab = win_find_tabpage(win), *save_curtab = curtab;
+
+  if (!win || !buf) {
+    return;
+  }
+
+  int result
+      = noautocmd
```

Comment:
could use `switch_win_noblock` always and instead `if (noautocmd) { block_autocmds(); }` and same on the other side.